proposals/excess_reserve_capacity.md

```diff
@@ -0,0 +1,49 @@
+
+* We can expand the scope of MachineDeployment controller to maintain an excess number of machines by introducing new parallel independent controller named *MachineTaint* controller. This will result in MCM to include Machine, MachineSet, MachineDeployment, MachienSafety, MachineTaint controllers. MachineTaint controller does not need to introduce any new CRD - analogy fits where taint-controller also resides into kube-controller-manager.
+
+* Only Job of MachineTaint controller will be:
+    * List all the Machines under each MachineDeployment.
+    * Put taints of [*noSchedule* and *noExecute*](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/) on latest MachineObjects.
+        * There should be an event-based informer mechanism where MachineTaintController gets to know about any Update/Delete/Create event of MachineObjects - in turn, maintains the *noSchedule* and *noExecute* taints on all the *latest* machines.
+        * Why latest machines?
+            * Whenever autoscaler decides to add new machines - essentially ScaleUp event - taints from the older machines are removed and newer machines get the taints. This way X number of Machines immediately becomes free for new pods to be scheduled.
+            * While ScaleDown event, autoscaler specifically mentions which machines should be deleted, and that should not bring any concerns. Though we will have to put proper label/annotation defined by autoscaler on taintedMachines, so that autoscaler does not consider the taintedMachines for deletion while scale-down.
+                * Annotation on tainted node: `"cluster-autoscaler.kubernetes.io/scale-down-disabled": "true"`
+* Implementation Details:
+    * Expect new **optional field** *ExcessReplicas* in `MachineDeployment.Spec`. MachineDeployment controller now adds both `Spec.Replicas` and `Spec.ExcessReplicas`[if provided], and considers that as a standard desiredReplicas.
+        * Current working of MCM will not be affected if ExcessReplicas field is kept nil.
+    * MachineController currently reads the *NodeObject* and sets the MachineConditions in MachineObject. Machine-controller will now also read the taints/labels from the MachineObject - and maintains it on the *NodeObject*.
+
+* We expect cluster-autoscaler to intelligently make use of the provided feature from MCM.
+    * CA gets the input of *min:max:excess* from Gardener. CA continues to set the `MachineDeployment.Spec.Replicas` as usual based on the application-workload.
+    * In addition, CA also sets the `MachieDeployment.Spec.ExcessReplicas` .
+    * Corner-case:
+        * CA should decrement the excessReplicas field accordingly when *desiredReplicas+excessReplicas* on MachineDeployment goes beyond *max*.
```

Comment:
Approach 1 looks workable.

However, one disadvantage I can see with this approach is that the granularity for excess capacity can only be a single machine. It cannot easily be made more generic to consider more fine-grained resource-based (cpu, memory, disk etc.) excess capacity.

Reply:
To me the excess capacity logic, should ideally be taken care of by the cluster autoscaler. Say even we implement this logic, and eventually the autoscaler would come up with algorithms to do similar things as excess reserve capacity - then this would create problems where two entities now would try to handle excess capacity. And then we would have to revert this change in the MCM's core logic again.

Although approach 1 sounds workable, it looks more like an temporary hack to me and I am not quite sure if it's the right way to go.

Reply:
I understand your point @prashanth, but don't know how to otherwise reach the goal. As for yours @amshu I thought the same, but then again the granularity is always a machine anyways.

I have a few questions:
1. How is the initial state (no load) handled? Are you starting with #min-#excess untainted and #excess tainted machines? I mean, we'd like to see a new machine the moment the load + desired excess reaches the overall (min) capacity.
2. In essence, the idea is to untaint a tainted machine the moment the cluster autoscaler requests a new one (fast) and start provisioning a new machine that will be forcefully tainted when available (slow). If even more machines are requested (no untained machines available), new machines will be provisioned, but not tainted on startup, right?
3. How will this work out "Though we will have to put proper label/annotation defined by autoscaler on taintedMachines, so that autoscaler does not consider the taintedMachines for deletion while scale-down."?

Reply:
As discussed over the call, we may utilize the PodPriority feature to reserve excess capacity in cluster. 
- After an initial investigation, PodPriority seems smooth enough to satisfy our use-case.

Though answering the questions to clarify the proposed approach:
 -> 1: Initially min+excess[tainted] would have been available. As we always want to maintain the min number of machines as specified by autoscaler.
-> 2: Yes, that is correct. Though I would further mention once the number of machines approaches MAX, we would reduce the tainted machines accordingly - eventually, cluster will have the max number of machines if required based on load- with no tainted machines.
-> 3: Autoscaler itself provides feature- where we can put a predefined annotation on the Node-object which will be excluded from the scale-down.
Hope that answers the queries :)

Reply:
Hmm... wasn't your idea to use low prio pods and move the topic into the seed cluster bootstrapper? Then we wouldn't have to do anything in the MCM.

Reply:
Yes, PodPriority/Low-priority-pods is still the preferred approach.

I felt like to answer the doubt for sake of answering related to other approach which was based on changes in MCM - though we would be not adapting this one. Sorry for the confusion :)

Reply:
I am confused too. Isn't PodPriority about using low priority pods to reserve capacity for additional control planes?

If so, then why do we need to worry about excess and tainted machines?

Reply:
Ah, thanks Hardik. You actually also wrote it: you just wanted to give answers to the questions raised to complete the discussion.

Then I take...

> After an initial investigation, PodPriority seems smooth enough to satisfy our use-case.

...as statement, that you tried it out and we can move the topic to the seed cluster bootstrapper and close this ticket as `won't-fix` for the time being?

Can you please elaborate what "seems smooth enough" mean? How fast do the low-prio pods get evicted and the new/actual ones get scheduled?